muzzle/src/test/java/io/opentelemetry/javaagent/tooling/muzzle/CacheProviderTest.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.tooling.muzzle;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.lang.ref.WeakReference;
+import java.net.URL;
+import java.net.URLClassLoader;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.pool.TypePool;
+import org.junit.jupiter.api.Test;
+
+class CacheProviderTest {
+
+  @Test
+  void keyBootstrapEquivalence() {
+    // ClassLoader loader = null;
+    int loaderHash = AgentCachingPoolStrategy.BOOTSTRAP_HASH;
+    WeakReference<ClassLoader> loaderRef = null;
+
+    AgentCachingPoolStrategy.TypeCacheKey key1 =
+        new AgentCachingPoolStrategy.TypeCacheKey(loaderHash, loaderRef, "foo");
+    AgentCachingPoolStrategy.TypeCacheKey key2 =
+        new AgentCachingPoolStrategy.TypeCacheKey(loaderHash, loaderRef, "foo");
+
+    assertThat(key1).hasSameHashCodeAs(key2);
+    assertThat(key1).isEqualTo(key2);
+  }
+
+  @Test
+  void keySameRefEquivalence() {
+    ClassLoader loader = newClassLoader();
+    int loaderHash = loader.hashCode();
+    WeakReference<ClassLoader> loaderRef = new WeakReference<>(loader);
+
+    AgentCachingPoolStrategy.TypeCacheKey key1 =
+        new AgentCachingPoolStrategy.TypeCacheKey(loaderHash, loaderRef, "foo");
+    AgentCachingPoolStrategy.TypeCacheKey key2 =
+        new AgentCachingPoolStrategy.TypeCacheKey(loaderHash, loaderRef, "foo");
+
+    assertThat(key1).hasSameHashCodeAs(key2);
+    assertThat(key1).isEqualTo(key2);
+    // ensures that loader isn't collected
+    assertThat(loader).isNotNull();
```

Comment:
loader can not be null at this point so, as it is, this assertion doesn't really make sense

Reply:
I think I get it now. The intention is to use the loader at the end of the method to ensure that it isn't collected before.